File: core/ws/cross_hub.go
package ws

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"strconv"
	"sync"
	"time"

	"hei-gin/config"
	"hei-gin/core/enums"

	"github.com/redis/go-redis/v9"
)

// Config helpers - read from config.C.WS with fallback defaults
func hbInterval() time.Duration {
	if config.C != nil && config.C.WS.HeartbeatInterval > 0 {
		return time.Duration(config.C.WS.HeartbeatInterval) * time.Second
	}
	return 15 * time.Second
}

func instTTL() time.Duration {
	if config.C != nil && config.C.WS.InstanceTTL > 0 {
		return time.Duration(config.C.WS.InstanceTTL) * time.Second
	}
	return 60 * time.Second
}

func staleClean() time.Duration {
	if config.C != nil && config.C.WS.StaleCleanInterval > 0 {
		return time.Duration(config.C.WS.StaleCleanInterval) * time.Minute
	}
	return 5 * time.Minute
}

func rlWindow() time.Duration {
	if config.C != nil && config.C.WS.RateLimitWindow > 0 {
		return time.Duration(config.C.WS.RateLimitWindow) * time.Second
	}
	return 10 * time.Second
}

func rlMax() int64 {
	if config.C != nil && config.C.WS.RateLimitMax > 0 {
		return int64(config.C.WS.RateLimitMax)
	}
	return 30
}

func dedupTTL() time.Duration {
	if config.C != nil && config.C.WS.DedupTTL > 0 {
		return time.Duration(config.C.WS.DedupTTL) * time.Second
	}
	return 30 * time.Second
}

func pollTO() time.Duration {
	if config.C != nil && config.C.WS.PollTimeout > 0 {
		return time.Duration(config.C.WS.PollTimeout) * time.Second
	}
	return 2 * time.Second
}

// crossInstanceMessage is the envelope for Redis list cross-instance messages.
type crossInstanceMessage struct {
	ToUserID   string          `json:"to_user_id"`
	ToUserType string          `json:"to_user_type"`
	Message    json.RawMessage `json:"message"`
	MessageID  string          `json:"message_id,omitempty"`
	Timestamp  int64           `json:"timestamp"`
}

// CrossHub wraps the local Hub with Redis-backed cross-instance IM delivery.
type CrossHub struct {
	local      *Hub
	rdb        *redis.Client
	instanceID string
	ctx        context.Context
	cancel     context.CancelFunc
	wg         sync.WaitGroup
	closeOnce  sync.Once
}

// NewCrossHub creates a CrossHub. If rdb is nil, runs in single-instance (passthrough) mode.
func NewCrossHub(local *Hub, rdb *redis.Client) *CrossHub {
	instanceID := strconv.FormatInt(config.C.Snowflake.Instance, 10)

	ctx, cancel := context.WithCancel(context.Background())
	ch := &CrossHub{
		local:      local,
		rdb:        rdb,
		instanceID: instanceID,
		ctx:        ctx,
		cancel:     cancel,
	}

	// Wire up lifecycle hooks to the local hub
	local.OnClientRegistered = ch.onClientRegistered
	local.OnClientUnregistered = ch.onClientUnregistered

	if rdb != nil {
		ch.wg.Add(3)
		go ch.pollLoop()
		go ch.heartbeatLoop()
		go ch.staleCleanupLoop()
		log.Printf("[CrossHub] Cross-instance mode enabled, instance=%s", instanceID)
	} else {
		log.Printf("[CrossHub] Redis not configured, running in single-instance mode")
	}

	return ch
}

// ─── Hub lifecycle hooks ──────────────────────────────────────────────

func (ch *CrossHub) onClientRegistered(c *Client) {
	ch.TrackConnection(c.UserID, c.UserType)
	ch.broadcastPresence(c.UserID, c.UserType, true)
	// Notify client to refresh unread count on reconnect
	switch c.UserType {
	case enums.LoginTypeBusiness:
		ch.local.SendToUser(c.UserID, Message{Type: MsgUnreadCount})
	case enums.LoginTypeConsumer:
		ch.local.SendToConsumer(c.UserID, Message{Type: MsgUnreadCount})
	}
}

func (ch *CrossHub) onClientUnregistered(c *Client) {
	// Remove this instance from user's connection set
	ch.UntrackConnection(c.UserID, c.UserType)
	// If user has no more connections on any instance, broadcast offline
	if !ch.IsUserOnlineAnywhere(c.UserID, c.UserType) {
		ch.broadcastPresence(c.UserID, c.UserType, false)
	}
}

// ─── Redis key helpers ────────────────────────────────────────────────

func (ch *CrossHub) userSetKey(userType enums.LoginTypeEnum, userID string) string {
	return "ws:user:" + string(userType) + ":" + userID
}

func (ch *CrossHub) msgListKey() string {
	return "ws:messages:" + ch.instanceID
}

func (ch *CrossHub) instanceKey() string {
	return "ws:instance:" + ch.instanceID
}

func (ch *CrossHub) rateLimitKey(userID string, userType enums.LoginTypeEnum) string {
	return "ws:ratelimit:" + string(userType) + ":" + userID
}

func (ch *CrossHub) dedupKey(messageID string) string {
	return "ws:dedup:" + messageID
}

// ─── Presence ─────────────────────────────────────────────────────────

// IsUserOnlineAnywhere checks if a user is connected to any instance.
func (ch *CrossHub) IsUserOnlineAnywhere(userID string, userType enums.LoginTypeEnum) bool {
	if ch.rdb == nil {
		return ch.local.isUserConnected(userID, userType)
	}
	key := ch.userSetKey(userType, userID)
	count, err := ch.rdb.SCard(ch.ctx, key).Result()
	if err != nil {
		return false
	}
	return count > 0
}

// IsUserOnlineLocally checks if a user is connected to this instance.
func (ch *CrossHub) IsUserOnlineLocally(userID string, userType enums.LoginTypeEnum) bool {
	return ch.local.isUserConnected(userID, userType)
}

// broadcastPresence sends presence update to conversation partners.
// For now, broadcasts to all connected clients of the same type.
// Can be optimized later by only notifying relevant conversation partners.
func (ch *CrossHub) broadcastPresence(userID string, userType enums.LoginTypeEnum, online bool) {
	msg := Message{
		Type: MsgPresence,
		Payload: PresencePayload{
			UserID:   userID,
			UserType: string(userType),
			Online:   online,
		},
	}
	ch.local.BroadcastAll(msg)
}

// TrackConnection records that a user is connected to this instance.
func (ch *CrossHub) TrackConnection(userID string, userType enums.LoginTypeEnum) {
	if ch.rdb == nil {
		return
	}
	key := ch.userSetKey(userType, userID)
	if err := ch.rdb.SAdd(ch.ctx, key, ch.instanceID).Err(); err != nil {
		log.Printf("[CrossHub] TrackConnection SAdd error: %v", err)
	}
	ch.rdb.Expire(ch.ctx, key, instTTL()+30*time.Second)
}

// UntrackConnection removes the instance from the user's instance set.
func (ch *CrossHub) UntrackConnection(userID string, userType enums.LoginTypeEnum) {
	if ch.rdb == nil {
		return
	}
	key := ch.userSetKey(userType, userID)
	if err := ch.rdb.SRem(ch.ctx, key, ch.instanceID).Err(); err != nil {
		log.Printf("[CrossHub] UntrackConnection SRem error: %v", err)
	}
}

// getTargetInstances returns all instance IDs where the user is connected.
func (ch *CrossHub) getTargetInstances(userID string, userType enums.LoginTypeEnum) []string {
	if ch.rdb == nil {
		return nil
	}
	key := ch.userSetKey(userType, userID)
	instances, err := ch.rdb.SMembers(ch.ctx, key).Result()
	if err != nil {
		log.Printf("[CrossHub] SMembers error: %v", err)
		return nil
	}
	return instances
}

// ─── Rate Limiting ────────────────────────────────────────────────────

// AllowMessage checks if a user is allowed to send a message (rate-limited).
func (ch *CrossHub) AllowMessage(userID string, userType enums.LoginTypeEnum) bool {
	if ch.rdb == nil {
		return true
	}
	key := ch.rateLimitKey(userID, userType)
	count, err := ch.rdb.Incr(ch.ctx, key).Result()
	if err != nil {
		return true // allow on error
	}
	if count == 1 {
		ch.rdb.Expire(ch.ctx, key, rlWindow())
	}
	return count <= rlMax()
}

// ─── Message Deduplication (cross-instance) ───────────────────────────

// markDelivered marks a message as delivered to prevent duplicate cross-instance delivery.
func (ch *CrossHub) markDelivered(messageID string) bool {
	if ch.rdb == nil || messageID == "" {
		return true
	}
	key := ch.dedupKey(messageID)
	// SETNX: only set if key doesn't exist; returns true if set successfully
	ok, err := ch.rdb.SetNX(ch.ctx, key, ch.instanceID, dedupTTL()).Result()
	if err != nil {
		return true // allow on error
	}
	return ok
}

// ─── Polling loop ─────────────────────────────────────────────────────

func (ch *CrossHub) pollLoop() {
	defer ch.wg.Done()
	key := ch.msgListKey()

	for {
		result, err := ch.rdb.BRPop(ch.ctx, pollTO(), key).Result()
		if err != nil {
			if err == context.Canceled || err == redis.Nil {
				select {
				case <-ch.ctx.Done():
					return
				default:
					continue
				}
			}
			log.Printf("[CrossHub] BRPop error: %v", err)
			continue
		}

		if len(result) < 2 {
			continue
		}
		ch.handleMessage(result[1])
	}
}

func (ch *CrossHub) handleMessage(payload string) {
	var xMsg crossInstanceMessage
	if err := json.Unmarshal([]byte(payload), &xMsg); err != nil {
		log.Printf("[CrossHub] Failed to unmarshal message: %v", err)
		return
	}

	// Dedup check: if this message ID was already processed, skip
	if xMsg.MessageID != "" {
		if !ch.markDelivered(xMsg.MessageID) {
			return // already processed
		}
	}

	var msg Message
	if err := json.Unmarshal(xMsg.Message, &msg); err != nil {
		log.Printf("[CrossHub] Failed to unmarshal inner message: %v", err)
		return
	}

	switch enums.LoginTypeEnum(xMsg.ToUserType) {
	case enums.LoginTypeBusiness:
		ch.local.SendToUser(xMsg.ToUserID, msg)
	case enums.LoginTypeConsumer:
		ch.local.SendToConsumer(xMsg.ToUserID, msg)
	}
}

// ─── Heartbeat ────────────────────────────────────────────────────────

func (ch *CrossHub) heartbeatLoop() {
	defer ch.wg.Done()
	ticker := time.NewTicker(hbInterval())
	defer ticker.Stop()

	ch.sendHeartbeat()

	for {
		select {
		case <-ch.ctx.Done():
			return
		case <-ticker.C:
			ch.sendHeartbeat()
		}
	}
}

func (ch *CrossHub) sendHeartbeat() {
	if ch.rdb == nil {
		return
	}
	key := ch.instanceKey()
	if err := ch.rdb.SetEx(ch.ctx, key, time.Now().UnixMilli(), instTTL()).Err(); err != nil {
		log.Printf("[CrossHub] Heartbeat error: %v", err)
	}
}

// ─── Stale Instance Cleanup ───────────────────────────────────────────

func (ch *CrossHub) staleCleanupLoop() {
	defer ch.wg.Done()
	ticker := time.NewTicker(staleClean())
	defer ticker.Stop()

	for {
		select {
		case <-ch.ctx.Done():
			return
		case <-ticker.C:
			ch.cleanStaleInstances()
		}
	}
}

// cleanStaleInstances removes stale instance entries from all user sets.
func (ch *CrossHub) cleanStaleInstances() {
	if ch.rdb == nil {
		return
	}

	// Find all user set keys
	iter := ch.rdb.Scan(ch.ctx, 0, "ws:user:*", 1000).Iterator()
	cleaned := 0

	for iter.Next(ch.ctx) {
		key := iter.Val()

		// Get all instances in this user set
		members, err := ch.rdb.SMembers(ch.ctx, key).Result()
		if err != nil {
			continue
		}

		for _, instID := range members {
			if instID == ch.instanceID {
				continue // skip our own instance
			}
			// Check if the instance is still alive
			instKey := "ws:instance:" + instID
			exists, err := ch.rdb.Exists(ch.ctx, instKey).Result()
			if err != nil {
				continue
			}
			if exists == 0 {
				// Instance is dead, remove from user set
				ch.rdb.SRem(ch.ctx, key, instID)
				cleaned++
			}
		}
	}

	if cleaned > 0 {
		log.Printf("[CrossHub] Cleaned %d stale instance references", cleaned)
	}
}

// ─── Public API ───────────────────────────────────────────────────────

// SendToUser delivers a message to a business user (local + cross-instance).
func (ch *CrossHub) SendToUser(userID string, msg Message, messageID ...string) {
	ch.local.SendToUser(userID, msg)
	if ch.rdb != nil {
		mid := ""
		if len(messageID) > 0 {
			mid = messageID[0]
		}
		ch.publishToRemote(userID, enums.LoginTypeBusiness, msg, mid)
	}
}

// SendToConsumer delivers a message to a consumer user (local + cross-instance).
func (ch *CrossHub) SendToConsumer(userID string, msg Message, messageID ...string) {
	ch.local.SendToConsumer(userID, msg)
	if ch.rdb != nil {
		mid := ""
		if len(messageID) > 0 {
			mid = messageID[0]
		}
		ch.publishToRemote(userID, enums.LoginTypeConsumer, msg, mid)
	}
}

// publishToRemote pushes a message to remote instances where the user is connected.
func (ch *CrossHub) publishToRemote(userID string, userType enums.LoginTypeEnum, msg Message, messageID string) {
	instances := ch.getTargetInstances(userID, userType)
	if len(instances) == 0 {
		return
	}

	rawMsg, err := json.Marshal(msg)
	if err != nil {
		return
	}

	xMsg := crossInstanceMessage{
		ToUserID:   userID,
		ToUserType: string(userType),
		Message:    rawMsg,
		MessageID:  messageID,
		Timestamp:  time.Now().UnixMilli(),
	}

	data, err := json.Marshal(xMsg)
	if err != nil {
		return
	}

	for _, instID := range instances {
		if instID == ch.instanceID {
			continue
		}
		listKey := "ws:messages:" + instID
		if err := ch.rdb.LPush(ch.ctx, listKey, data).Err(); err != nil {
			log.Printf("[CrossHub] LPush to %s error: %v", instID, err)
		}
	}
}

// HandleWebSocket upgrades an HTTP connection to WebSocket with cross-instance tracking.
func (ch *CrossHub) HandleWebSocket(w http.ResponseWriter, r *http.Request, userID string, userType enums.LoginTypeEnum) {
	ch.local.HandleWebSocket(w, r, userID, userType)
}

// OnlineCount returns locally connected client count.
func (ch *CrossHub) OnlineCount() int {
	return ch.local.OnlineCount()
}

// Close shuts down the CrossHub gracefully.
func (ch *CrossHub) Close() {
	if ch == nil {
		return
	}
	ch.closeOnce.Do(func() {
		ch.cancel()
		if ch.rdb != nil {
			ch.rdb.Del(ch.ctx, ch.instanceKey())
		}
		ch.wg.Wait()
	})
}

File: plugins/plugin_im/ws/cross_hub.py
"""
CrossHub — Redis-backed cross-instance WebSocket message delivery.

Mirrors hei-gin's ``plugins/plugin-im/ws/cross_hub.go``.
"""

from __future__ import annotations

import asyncio
import json
import logging
import time

from .hub import Hub
from .client import Client
from .message import (
    Message, MsgUnreadCount, MsgPresence, PresencePayload,
)
from .config import ws_cfg
from sdk.infra.db import get_redis
from sdk.infra.eventbus import publish as event_publish

logger = logging.getLogger(__name__)


class CrossHub:
    """Wraps a local Hub with Redis-backed cross-instance IM delivery."""

    def __init__(self, local: Hub) -> None:
        self.local = local
        self._rdb = get_redis()
        self._instance_id = "0"
        self._running = True
        self._tasks: set[asyncio.Task] = set()

        # Try to get instance ID from settings
        try:
            from sdk.config.settings import settings
            self._instance_id = str(getattr(settings, "snowflake", {}).get("instance", 0) or 0)
        except Exception:
            self._instance_id = "0"

        # Register hub lifecycle hooks
        local.on_client_registered = self._on_client_registered
        local.on_client_unregistered = self._on_client_unregistered

        if self._rdb:
            logger.info("CrossHub cross-instance mode enabled, instance=%s", self._instance_id)

    # ── Lifecycle hooks ──────────────────────────────────────────────

    def _on_client_registered(self, client: Client) -> None:
        self.create_task(self._track_connection(client))
        self.create_task(self._broadcast_presence(client.user_id, client.user_type, True))
        self.create_task(self._send_unread_on_connect(client))
        self.create_task(self._publish_event("user:connected", client))

    def _on_client_unregistered(self, client: Client) -> None:
        self.create_task(self._untrack_connection(client))
        self.create_task(self._check_presence(client))
        self.create_task(self._publish_event("user:disconnected", client))

    def create_task(self, coro) -> asyncio.Task | None:
        if not self._running:
            return None
        try:
            loop = asyncio.get_running_loop()
        except RuntimeError:
            return None
        task = loop.create_task(coro)
        self._tasks.add(task)
        task.add_done_callback(self._tasks.discard)
        return task

    def refresh_redis(self) -> None:
        self._rdb = get_redis()
        if self._rdb:
            logger.info("CrossHub cross-instance mode enabled, instance=%s", self._instance_id)

    async def _publish_event(self, topic: str, client: Client) -> None:
        try:
            await event_publish(topic, client=client)
        except Exception:
            logger.exception("Failed to publish event %s", topic)

    async def _send_unread_on_connect(self, client: Client) -> None:
        msg = Message(type=MsgUnreadCount)
        if client.user_type == "BUSINESS":
            await self.local.send_to_user(client.user_id, msg)
        else:
            await self.local.send_to_consumer(client.user_id, msg)

    async def _check_presence(self, client: Client) -> None:
        if not await self._is_user_online_anywhere(client.user_id, client.user_type):
            await self._broadcast_presence(client.user_id, client.user_type, False)

    # ── Redis helpers ────────────────────────────────────────────────

    def _user_set_key(self, user_type: str, user_id: str) -> str:
        return f"ws:user:{user_type}:{user_id}"

    def _instance_key(self) -> str:
        return f"ws:instance:{self._instance_id}"

    def _msg_list_key(self) -> str:
        return f"ws:messages:{self._instance_id}"

    def _dedup_key(self, message_id: str) -> str:
        return f"ws:dedup:{message_id}"

    def _rate_limit_key(self, user_id: str, user_type: str) -> str:
        return f"ws:ratelimit:{user_type}:{user_id}"

    # ── Connection tracking ──────────────────────────────────────────

    async def _track_connection(self, client: Client) -> None:
        if not self._rdb:
            return
        key = self._user_set_key(client.user_type, client.user_id)
        await self._rdb.sadd(key, self._instance_id)
        await self._rdb.expire(key, ws_cfg.instance_ttl * 2)

        inst_key = self._instance_key()
        await self._rdb.setex(inst_key, ws_cfg.instance_ttl, str(time.time()))

    async def _untrack_connection(self, client: Client) -> None:
        if not self._rdb:
            return
        key = self._user_set_key(client.user_type, client.user_id)
        await self._rdb.srem(key, self._instance_id)

    async def _get_target_instances(self, user_id: str, user_type: str) -> list[str]:
        """Get all instance IDs where a user is connected."""
        if not self._rdb:
            return []
        try:
            members = await self._rdb.smembers(self._user_set_key(user_type, user_id))
            return [m.decode() if isinstance(m, bytes) else m for m in members]
        except Exception:
            return []

    async def _is_user_online_anywhere(self, user_id: str, user_type: str) -> bool:
        if not self._rdb:
            return self.local.is_user_connected(user_id, user_type)
        try:
            members = await self._rdb.smembers(self._user_set_key(user_type, user_id))
            return len(members) > 0
        except Exception:
            return False

    # ── Presence ─────────────────────────────────────────────────────

    async def _broadcast_presence(self, user_id: str, user_type: str, online: bool) -> None:
        msg = Message(
            type=MsgPresence,
            payload=PresencePayload(
                user_id=user_id, user_type=user_type, online=online,
            ).__dict__,
        )
        await self.local.broadcast_business(msg)
        await self.local.broadcast_consumers(msg)

    # ── Rate limiting ────────────────────────────────────────────────

    async def allow_message(self, sender_id: str, sender_type: str) -> bool:
        if not self._rdb:
            return True
        key = self._rate_limit_key(sender_id, sender_type)
        try:
            current = await self._rdb.incr(key)
            if current == 1:
                await self._rdb.expire(key, ws_cfg.rate_limit_window)
            return current <= ws_cfg.rate_limit_max
        except Exception:
            return True

    # ── Dedup ────────────────────────────────────────────────────────

    async def is_dedup(self, message_id: str) -> bool:
        if not self._rdb or not message_id:
            return False
        key = self._dedup_key(message_id)
        try:
            existed = await self._rdb.setnx(key, "1")
            if existed:
                await self._rdb.expire(key, ws_cfg.dedup_ttl)
                return False
            return True
        except Exception:
            return False

    # ── Cross-instance delivery ──────────────────────────────────────

    async def _publish_to_remote(
        self, user_id: str, user_type: str, msg: Message, message_id: str = ""
    ) -> None:
        if not self._rdb:
            return
        instances = await self._get_target_instances(user_id, user_type)
        if not instances:
            return

        raw_msg = json.dumps(msg.to_dict(), ensure_ascii=False)
        x_msg = json.dumps({
            "to_user_id": user_id,
            "to_user_type": user_type,
            "message": raw_msg,
            "message_id": message_id,
            "timestamp": int(time.time() * 1000),
        }, ensure_ascii=False)

        for inst_id in instances:
            if inst_id == self._instance_id:
                continue
            list_key = f"ws:messages:{inst_id}"
            try:
                await self._rdb.lpush(list_key, x_msg)
                await self._rdb.expire(list_key, 300)  # 5 min TTL
            except Exception:
                logger.exception("Failed to push to instance %s", inst_id)

    # ── Public API ───────────────────────────────────────────────────

    async def send_to_user(self, user_id: str, msg: Message, message_id: str = "") -> None:
        await self.local.send_to_user(user_id, msg)
        await self._publish_to_remote(user_id, "BUSINESS", msg, message_id)

    async def send_to_users(self, user_ids: list[str], msg: Message) -> None:
        await self.local.send_to_users(user_ids, msg)
        for uid in user_ids:
            await self._publish_to_remote(uid, "BUSINESS", msg)

    async def send_to_consumer(self, user_id: str, msg: Message, message_id: str = "") -> None:
        await self.local.send_to_consumer(user_id, msg)
        await self._publish_to_remote(user_id, "CONSUMER", msg, message_id)

    async def send_to_consumers(self, user_ids: list[str], msg: Message) -> None:
        await self.local.send_to_consumers(user_ids, msg)
        for uid in user_ids:
            await self._publish_to_remote(uid, "CONSUMER", msg)

    async def broadcast_all(self, msg: Message) -> None:
        await self.local.broadcast_all(msg)

    async def broadcast_business(self, msg: Message) -> None:
        await self.local.broadcast_business(msg)

    async def broadcast_consumers(self, msg: Message) -> None:
        await self.local.broadcast_consumers(msg)

    def online_count(self) -> int:
        return self.local.online_count()

    # ── Poll loop ────────────────────────────────────────────────────

    async def start_poll_loop(self) -> None:
        """Poll for cross-instance messages intended for this instance."""
        if not self._rdb:
            return
        while self._running:
            try:
                result = await self._rdb.brpop(self._msg_list_key(), timeout=ws_cfg.poll_timeout)
                if result is None:
                    continue
                _, raw = result
                data = json.loads(raw)
                msg_data = json.loads(data["message"])
                msg = Message(**msg_data)

                to_type = data.get("to_user_type", "BUSINESS")
                to_id = data.get("to_user_id")

                if to_type == "BUSINESS":
                    await self.local.send_to_user(to_id, msg)
                else:
                    await self.local.send_to_consumer(to_id, msg)
            except Exception:
                if self._running:
                    logger.exception("CrossHub poll error")

    # ── Heartbeat ────────────────────────────────────────────────────

    async def start_heartbeat(self) -> None:
        """Periodically refresh the instance key in Redis."""
        if not self._rdb:
            return
        interval = max(ws_cfg.heartbeat_interval, 15)
        while self._running:
            await asyncio.sleep(interval)
            try:
                inst_key = self._instance_key()
                await self._rdb.setex(inst_key, ws_cfg.instance_ttl, str(time.time()))
            except Exception:
                logger.exception("CrossHub heartbeat error")

    # ── Stale Instance Cleanup ───────────────────────────────────────

    async def start_stale_cleanup(self) -> None:
        """Periodically remove stale instance references from Redis user sets."""
        if not self._rdb:
            return
        interval = max(ws_cfg.stale_clean_interval, 1) * 60
        while self._running:
            await asyncio.sleep(interval)
            try:
                await self._clean_stale_instances()
            except Exception:
                logger.exception("CrossHub stale cleanup error")

    async def _clean_stale_instances(self) -> None:
        """Scan Redis for user sets and remove references to dead instances."""
        cursor = 0
        cleaned = 0
        pattern = "ws:user:*"
        while True:
            cursor, keys = await self._rdb.scan(cursor, match=pattern, count=1000)
            for key in keys:
                members = await self._rdb.smembers(key)
                for inst_id in members:
                    inst_id_str = inst_id.decode() if isinstance(inst_id, bytes) else inst_id
                    if inst_id_str == self._instance_id:
                        continue
                    inst_key = f"ws:instance:{inst_id_str}"
                    exists = await self._rdb.exists(inst_key)
                    if not exists:
                        await self._rdb.srem(key, inst_id)
                        cleaned += 1
            if cursor == 0:
                break
        if cleaned > 0:
            logger.info("Cleaned %d stale instance references", cleaned)

    # ── Message List Cleanup ─────────────────────────────────────────

    async def start_msg_list_cleanup(self) -> None:
        """Periodically trim the message list and refresh TTL."""
        if not self._rdb:
            return
        while self._running:
            await asyncio.sleep(300)  # 5 minutes
            try:
                key = self._msg_list_key()
                await self._rdb.ltrim(key, -1000, -1)
                await self._rdb.expire(key, 300)
            except Exception:
                logger.exception("CrossHub msg list cleanup error")

    # ── Close ────────────────────────────────────────────────────────

    async def close(self) -> None:
        self._running = False
        self.local.on_client_registered = None
        self.local.on_client_unregistered = None
        if self._rdb:
            try:
                await self._rdb.delete(self._instance_key())
                await self._rdb.delete(self._msg_list_key())
            except Exception:
                pass
        await self.local.close()
        if self._tasks:
            for task in list(self._tasks):
                task.cancel()
            await asyncio.gather(*list(self._tasks), return_exceptions=True)
            self._tasks.clear()
